solarforecastarbiter/validation/quality_mapping.py

```diff
@@ -0,0 +1,133 @@
+"""
+Define constant mappings between bit-mask values and understandable quality
+flags
+"""
+from functools import wraps
+
+
+# The quality_flag field in MySQL is currently limited to 1 << 15;
+# fields beyond 1 << 15 will require a change in the MySQL datatype
+# for the quality_flag column. The mapping from description to bitmask
+# is versioned so that future addtions or removals are backwards compatible
+# without rerunning the validation on all data.
+# DO NOT MODIFY THE VALUES OF THE _BITMASK_DESCRIPTION_DICT instead,
+# add a increment the key and add a new value tuple. The VERSION
+# IDENTIFIER 0 - 2 must remain in their current positions. Versions 7
+# and up will require another identifier bit to be determined at that
+# time.  The version identifier also serves to mark data in the
+# database as validated. The tuples are (description, bit mask, flag class)
+_BITMASK_DESCRIPTION_DICT = {1: (
+    # start with 1 to distinguish validated vs not in DB
+    ('OK', 0, ''),
+    ('USER FLAGGED', 1 << 0, 'bad'),
+    ('VERSION IDENTIFIER 0', 1 << 1, ''),
+    ('VERSION IDENTIFIER 1', 1 << 2, ''),
+    ('VERSION IDENTIFIER 2', 1 << 3, ''),
+    ('NIGHTTIME', 1 << 4, 'informational'),
+    ('CLOUDY', 1 << 5, 'informational'),
+    ('SHADED', 1 << 6, 'informational'),
+    ('UNEVEN FREQUENCY', 1 << 7, 'bad'),
+    ('LIMITS EXCEEDED', 1 << 8, 'bad'),
+    ('CLEARSKY EXCEEDED', 1 << 9, 'bad'),
+    ('STALE VALUES', 1 << 10, 'bad'),
+    ('INTERPOLATED VALUES', 1 << 11, 'bad'),
+    ('CLIPPED VALUES', 1 << 12, 'bad'),
+    ('INCONSISTENT IRRADIANCE COMPONENTS', 1 << 13, 'bad'),
+    ('RESERVED 0', 1 << 14, ''),  # available for new flag
+    ('RESERVED 1', 1 << 15, ''),  # available for new flag
```

Comment:
looks like we're almost out of options already. should we go ahead and change MYSQL?

Reply:
Adding another field would add another byte to every observation value row. It isn't much and may not have any effect on performance, but I would rather not until there are clear uses for other flags. Also, the current flags can be  rename, removed, or combined if that makes sense to do.

Reply:
Would it be difficult to change in the future? Sounds like it could require a big table rebuild but I have no idea what I'm talking about here.

Reply:
It would require an ``ALTER TABLE`` command that may take an hour to run once the table is huge. Basically, if we need to take the reserved spots and keep the others in the next few months, we should go ahead and  add the extra possibilities. If we don't need more soon, I'm not sure how likely it is that we ever will.